substrate/frame/staking-async/runtimes/rc/src/lib.rs

```diff
@@ -606,19 +606,129 @@ enum RcClientCalls<AccountId> {
+		maybe_max_steps: Option<u32>,
+	) -> Result<Vec<Xcm<()>>, SendError> {
+		let mut chunk_size = report.validator_points.len();
+		let mut steps = 0;
+
+		loop {
+			let current_reports = report.clone().split(chunk_size);
+
+			// the first report is the heaviest, the last one might be smaller.
+			let first_report = if let Some(r) = current_reports.first() {
+				r
+			} else {
+				log::debug!(target: "runtime::ah-client", "📨 unexpected: no reports to send");
+				return Ok(vec![]);
+			};
+
+			log::debug!(
+				target: "runtime::ah-client",
+				"📨 step: {:?}, chunk_size: {:?}, report_size: {:?}",
+				steps,
+				chunk_size,
+				first_report.encoded_size(),
+			);
+
+			let message = Self::session_message_from_report(first_report.clone());
+			match <T as SendXcm>::validate(&mut Some(destination.clone()), &mut Some(message)) {
+				Ok((_ticket, price)) => {
+					log::debug!(target: "runtime::ah-client", "📨 validated, price: {:?}", price);
+					return Ok(current_reports
+						.into_iter()
+						.map(Self::session_message_from_report)
+						.collect::<Vec<_>>());
+				},
+				Err(SendError::ExceedsMaxMessageSize) => {
+					log::debug!(target: "runtime::ah-client", "📨 ExceedsMaxMessageSize -- reducing chunk_size");
+					chunk_size = chunk_size.saturating_div(2);
+					steps += 1;
+					if maybe_max_steps.map_or(false, |max_steps| steps > max_steps) {
+						log::error!(target: "runtime::ah-client", "📨 Exceeded max steps");
+						return Err(SendError::ExceedsMaxMessageSize);
+					} else {
+						// try again with the new `chunk_size`
+						continue;
+					}
+				},
+				Err(other) => {
+					log::error!(target: "runtime::ah-client", "📨 other error -- cannot send XCM: {:?}", other);
+					return Err(other);
+				},
+			}
+		}
+	}
+
+	fn session_message_from_report(report: rc_client::SessionReport<AccountId>) -> Xcm<()> {
+		Xcm(vec![
 			Instruction::UnpaidExecution {
 				weight_limit: WeightLimit::Unlimited,
 				check_origin: None,
 			},
-			Self::mk_asset_hub_call(RcClientCalls::RelaySessionReport(session_report)),
-		]);
-		if let Err(err) = send_xcm::<T>(AssetHubNextLocation::get(), message) {
-			log::error!(target: "runtime", "Failed to send relay session report message: {:?}", err);
+			Self::mk_asset_hub_call(RcClientCalls::RelaySessionReport(report)),
+		])
+	}
+
+	fn mk_asset_hub_call(
+		call: RcClientCalls<<Self as ah_client::SendToAssetHub>::AccountId>,
+	) -> Instruction<()> {
+		Instruction::Transact {
+			origin_kind: OriginKind::Superuser,
+			fallback_max_weight: None,
+			call: AssetHubRuntimePallets::RcClient(call).encode().into(),
+		}
+	}
+}
+
+impl<T: SendXcm, AssetHubId: Get<u32>> ah_client::SendToAssetHub for XcmToAssetHub<T, AssetHubId> {
+	type AccountId = AccountId;
+
+	fn relay_session_report(mut session_report: rc_client::SessionReport<Self::AccountId>) {
+		// add some fake data to the session report to make it bigger
+		session_report.validator_points.extend(
+			(0..1000)
+				.into_iter()
```

Comment:
Afaict this is not tests/mock code, why inflate it with dummy data for production code?

Reply:
It is a test runtime as noted above.